File: lab02/Factor.py
from math import isqrt
from collections import defaultdict
from typing import List, Dict

def factorization_string(n: int) -> str:
    factorization_result = factorization(n)

    result_string = " x ".join(
        f"{key}^{value}" if value > 1 else str(key)
        for key, value in factorization_result.items()
    )

    return result_string

def factorization(n: int) -> Dict[int, int]:
    factors = factor(n)

    grouped_factors = defaultdict(int)
    for currentFactor in factors:
        grouped_factors[currentFactor] += 1

    return dict(grouped_factors)

def factor(n: int) -> List[int]:
    result = []

    # Remove powers of 2, 3, 5
    for prime in [2, 3, 5]:
        n = remove_factors(n, result, prime)

    d = 0

    # Trial division with selected forms of d
    square_root = isqrt(n)
    while n > 1 and d <= square_root:
        forms_of_d = [1, 7, 11, 13, 17, 19, 23, 29]

        for current_residue in forms_of_d:
            current_divisor = d + current_residue

            if current_divisor != 1 and n % current_divisor == 0:
                n = remove_factors(n, result, current_divisor)

        d += 30  # Move to the next set of forms

    if n > 1:
        result.append(n)

    return result

def remove_factors(n: int, factors: List[int], prime: int) -> int:
    while n % prime == 0:
        factors.append(prime)
        n //= prime

    return n

def main():
    n = int(input("Enter n: "))
    factorizationString = factorization_string(n)
    print(factorizationString)

main()

File: lab02/cpp/Factor.cpp
#include <iostream>
#include <vector>
#include <map>
#include <cmath>
#include <string>
#include <sstream>
#include <algorithm>
#include <iterator>
#include <boost/multiprecision/cpp_int.hpp>

using BigInteger = boost::multiprecision::cpp_int;

BigInteger removeFactors(BigInteger n, std::vector<BigInteger>& factors, const BigInteger& prime) {
    while (n % prime == 0) {
        factors.push_back(prime);
        n /= prime;
    }

    return n;
}

std::vector<BigInteger> factor(BigInteger n) {
    std::vector<BigInteger> result;

    // Remove powers of 2, 3, 5
    for (const BigInteger& prime : {2, 3, 5}) {
        n = removeFactors(n, result, prime);
    }

    BigInteger d = 0;

    // Trial division with selected forms of d
    BigInteger squareRoot = sqrt(n);
    while (n > 1 && d <= squareRoot) {
        const std::vector<BigInteger> formsOfD = {1, 7, 11, 13, 17, 19, 23, 29};

        for (const BigInteger& currentResidue : formsOfD) {
            BigInteger currentDivisor = d + currentResidue;

            if (currentDivisor != 1 && n % currentDivisor == 0) {
                n = removeFactors(n, result, currentDivisor);
            }
        }

        d += 30;
    }

    if (n > 1) {
        result.push_back(n);
    }

    return result;
}

std::map<BigInteger, int> factorization(BigInteger n) {
    std::vector<BigInteger> factors = factor(n);
    std::map<BigInteger, int> groupedFactors;

    for (const BigInteger& factor : factors) {
        groupedFactors[factor]++;
    }

    return groupedFactors;
}

std::string factorizationString(const BigInteger& n) {
    std::map<BigInteger, int> factorizationResult = factorization(n);

    std::ostringstream resultStream;
    bool first = true;

    for (const auto& [key, value] : factorizationResult) {
        if (!first) {
            resultStream << " x ";
        }
        first = false;
        if (value > 1) {
            resultStream << key << "^" << value;
        } else {
            resultStream << key;
        }
    }

    return resultStream.str();
}



int main() {
    BigInteger n;
    std::cout << "Enter a number: ";
    std::cin >> n;

    std::string result = factorizationString(n);
    std::cout << "Factorization: " << result << std::endl;

    return 0;
}
